src/Namecheap/Provider.php

```diff
@@ -0,0 +1,502 @@
+        } catch (\Throwable $e) {
+            $this->handleException($e, $params);
+        }
+    }
+
+    /**
+     * @param  DomainInfoParams  $params
+     *
+     * @return DomainResult
+     */
+    public function getInfo(DomainInfoParams $params): DomainResult
+    {
+        $domainName = Utils::getDomain(
+            Utils::normalizeSld($params->sld),
+            Utils::normalizeTld($params->tld)
+        );
+
+        try {
+            return $this->_getInfo($domainName, 'Domain data obtained');
+        } catch (\Throwable $e) {
+            $this->handleException($e, $params);
+        }
+    }
+
+    /**
+     * @param  UpdateDomainContactParams  $params
+     *
+     * @return ContactResult
+     */
+    public function updateRegistrantContact(UpdateDomainContactParams $params): ContactResult
+    {
+        try {
+            $contact = $this->api()
+                ->updateRegistrantContact(
+                    Utils::getDomain(
+                        Utils::normalizeSld($params->sld),
+                        Utils::normalizeTld($params->tld)
+                    ),
+                    $params->contact
+                );
+
+            return ContactResult::create($contact);
+        } catch (\Throwable $e) {
+            $this->handleException($e, $params);
+        }
+    }
+
+    /**
+     * @param  UpdateNameserversParams  $params
+     *
+     * @return NameserversResult
+     */
+    public function updateNameservers(UpdateNameserversParams $params): NameserversResult
+    {
+        $sld = Utils::normalizeSld($params->sld);
+        $tld = Utils::normalizeTld($params->tld);
+
+        $domainName = Utils::getDomain($sld, $tld);
+
+        $nameServers = $this->prepareNameservers($params, 'ns');
+
+        try {
+            $result = $this->api()->updateNameservers(
+                $sld,
+                $tld,
+                $nameServers,
+            );
+
+            return NameserversResult::create($result)
+                ->setMessage(sprintf('Name servers for %s domain were updated!', $domainName));
+        } catch (\Throwable $e) {
+            $this->handleException($e, $params);
+        }
+    }
+
+    /**
+     * @param  LockParams  $params
+     *
+     * @return DomainResult
+     */
+    public function setLock(LockParams $params): DomainResult
+    {
+        $domainName = Utils::getDomain($params->sld, $params->tld);
+        $lock = !!$params->lock;
+
+        try {
+            if (in_array($params->tld, NamecheapApi::NO_LOCK_TLDS)) {
+                throw $this->errorResult(
+                    sprintf('Domain %s does not support to change Registrar Lock.', $domainName),
+                    $params
+                );
+            }
+
+            $currentLockStatus = $this->api()->getRegistrarLockStatus($domainName);
+            if (!$lock && !$currentLockStatus) {
+                throw $this->errorResult(sprintf('Domain %s already unlocked', $domainName), $params);
```

Comment:
Because provision function errors logs will typically lead to more administrative work for web hosts, wherever possible please just return a successful result if the domain is already in the desired state (e.g., already locked)